File: src/api/empleados.ts
import { callCommand } from './client'
import type {
  Audit,
  CivilDate,
  Decimal4,
  ListQuery,
  LookupItem,
  Money,
  PagedResult,
  RowVersion,
  Uuid,
} from './types'

/** See `docs/11-contratos-tauri.md` §5.7. */

export type FrecuenciaPago = 'Diario' | 'Semanal' | 'Quincenal' | 'Mensual'

export interface EmpleadoFiltro {
  texto?: string
  /** Absent means everyone; the list screen sends `true`. */
  activo?: boolean
  cargo?: string
}

export interface EmpleadoInput {
  nombre: string
  dni: string | null
  cargo: string | null
  sueldoBase: Money
  pagoFrecuencia: FrecuenciaPago
  tarifaDiaria: Money
  multiplicadorSabado: Decimal4
  multiplicadorDomingo: Decimal4
  multiplicadorFeriado: Decimal4
  email: string | null
  telefono: string | null
  fechaIngreso: CivilDate
  fechaEgreso: CivilDate | null
  activo: boolean
}

export interface EmpleadoListItem {
  id: Uuid
  nombre: string
  dni: string | null
  cargo: string | null
  tarifaDiaria: Money
  sueldoBase: Money
  pagoFrecuencia: FrecuenciaPago
  email: string | null
  telefono: string | null
  fechaIngreso: CivilDate
  fechaEgreso: CivilDate | null
  activo: boolean
  rowVersion: RowVersion
}

export interface EmpleadoDetalle {
  id: Uuid
  nombre: string
  dni: string | null
  cargo: string | null
  sueldoBase: Money
  pagoFrecuencia: FrecuenciaPago
  tarifaDiaria: Money
  /** What the rate would be if derived from the salary, offered by the form. */
  tarifaDiariaSugerida: Money
  multiplicadorSabado: Decimal4
  multiplicadorDomingo: Decimal4
  multiplicadorFeriado: Decimal4
  email: string | null
  telefono: string | null
  fechaIngreso: CivilDate
  fechaEgreso: CivilDate | null
  activo: boolean
  /** False when settlements, attendance or movements depend on the employee. */
  puedeEliminarse: boolean
  audit: Audit
}

export const EMPLEADOS_SORTABLE = [
  'nombre',
  'cargo',
  'tarifaDiaria',
  'sueldoBase',
  'fechaIngreso',
] as const

export function listEmpleados(
  query: ListQuery<EmpleadoFiltro>,
): Promise<PagedResult<EmpleadoListItem>> {
  return callCommand('empleados_list', { query })
}

export function getEmpleado(id: Uuid): Promise<EmpleadoDetalle> {
  return callCommand('empleados_get', { id })
}

export function lookupEmpleados(
  soloActivos?: boolean,
  texto?: string,
  limite?: number,
): Promise<LookupItem[]> {
  return callCommand('empleados_lookup', { soloActivos, texto, limite })
}

/** The roles already in use, so the filter offers what exists instead of a free field. */
export function cargosEmpleados(): Promise<string[]> {
  return callCommand('empleados_cargos', {})
}

export function createEmpleado(dto: EmpleadoInput): Promise<EmpleadoDetalle> {
  return callCommand('empleados_create', { dto })
}

export function updateEmpleado(
  id: Uuid,
  dto: EmpleadoInput,
  rowVersion: RowVersion,
): Promise<EmpleadoDetalle> {
  return callCommand('empleados_update', { id, dto, rowVersion })
}

export function deleteEmpleado(id: Uuid, rowVersion: RowVersion): Promise<void> {
  return callCommand('empleados_delete', { id, rowVersion })
}

File: src-tauri/src/commands/empleados.rs
//! Commands of `empleados`. See `docs/11-contratos-tauri.md` §5.7.

use eo_application::dtos::common::{ListQuery, LookupItem};
use eo_application::dtos::empleados::{
    EmpleadoDetalle, EmpleadoFiltroDto, EmpleadoInput, EmpleadoListItem,
};
use eo_application::PagedResult;
use tauri::State;
use uuid::Uuid;

use crate::error::{handle, ApiResult};
use crate::state::AppState;

#[tauri::command]
pub async fn empleados_list(
    state: State<'_, AppState>,
    query: ListQuery<EmpleadoFiltroDto>,
) -> ApiResult<PagedResult<EmpleadoListItem>> {
    let outcome = match state.services() {
        Ok(services) => services.empleados.list(query).await,
        Err(e) => Err(e),
    };
    handle("empleados_list", outcome)
}

#[tauri::command]
pub async fn empleados_get(state: State<'_, AppState>, id: Uuid) -> ApiResult<EmpleadoDetalle> {
    let outcome = match state.services() {
        Ok(services) => services.empleados.get(id).await,
        Err(e) => Err(e),
    };
    handle("empleados_get", outcome)
}

#[tauri::command]
pub async fn empleados_lookup(
    state: State<'_, AppState>,
    solo_activos: Option<bool>,
    texto: Option<String>,
    limite: Option<u64>,
) -> ApiResult<Vec<LookupItem>> {
    let outcome = match state.services() {
        Ok(services) => services.empleados.lookup(solo_activos, texto, limite).await,
        Err(e) => Err(e),
    };
    handle("empleados_lookup", outcome)
}

/// The distinct roles already in use, so the filter offers what exists instead of a free field.
#[tauri::command]
pub async fn empleados_cargos(state: State<'_, AppState>) -> ApiResult<Vec<String>> {
    let outcome = match state.services() {
        Ok(services) => services.empleados.cargos().await,
        Err(e) => Err(e),
    };
    handle("empleados_cargos", outcome)
}

#[tauri::command]
pub async fn empleados_create(
    state: State<'_, AppState>,
    dto: EmpleadoInput,
) -> ApiResult<EmpleadoDetalle> {
    let outcome = match state.services() {
        Ok(services) => services.empleados.create(dto).await,
        Err(e) => Err(e),
    };
    handle("empleados_create", outcome)
}

#[tauri::command]
pub async fn empleados_update(
    state: State<'_, AppState>,
    id: Uuid,
    dto: EmpleadoInput,
    row_version: String,
) -> ApiResult<EmpleadoDetalle> {
    let outcome = match state.services() {
        Ok(services) => services.empleados.update(id, dto, &row_version).await,
        Err(e) => Err(e),
    };
    handle("empleados_update", outcome)
}

#[tauri::command]
pub async fn empleados_delete(
    state: State<'_, AppState>,
    id: Uuid,
    row_version: String,
) -> ApiResult<()> {
    let outcome = match state.services() {
        Ok(services) => services.empleados.delete(id, &row_version).await,
        Err(e) => Err(e),
    };
    handle("empleados_delete", outcome)
}
